### PR TITLE
I/5862: Disable text transformations inside code blocks

### PR DESCRIPTION
**Pull request by panr**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Disabled text transformation inside code blocks. Closes ckeditor/ckeditor5#5862.

### CODE REVIEW

**Comment by jodator** [comment]
> * Tension: What's the best way for implementing `this._disableStack`

Looking at what we have already and what is easiest to use

1. `EmiterMixin` and `ObservableMixin` - It's OK to add `if( !this._disabledStack) ` check directly in mixin methods. Or some initializer method that will be called.
2. The other case I've found is `ElementApiMixin` which throws if a required property is not set. It does that because it must be mixed with proper editor class. So this case is not for `ForceDisablableMixin`

> Should `ObservableMixin` implement this

I don't think so - they don't know about each other and shouldn't know.

The name `ForceDisabledMixin` - it's pretty clear for me what it does (adds `forceDisabled()` and `clearForceDisabled()` methods ).

:thinking: Another name that I could think of is `DisableableMixn` (as in `ObservableMixin`) - which is kinda funny, so no :D We also do not follow any pattern here (like using `-able` suffix) so again I don't have a better name.

TL;DR:

1. `this._forceDisabled` set by methods (no throw) similar to `EmitterMixin`.
2. The name is OK for me.

**Comment by Reinmar** [comment]
What do you think about not exporting this to a mixin but rather keeping duplicated in the base `Command` and `Plugin` classes? Cons: duplication. Pros: no problem how this mixin should work. I'd consider this option as a possible middle ground between full flavoured completely reusable mixin and lots of code duplication (when every plugin and command would implement this manually).

**Comment by Reinmar** [comment]
Cause: "Duplication is better than the wrong abstraction".

**Comment by Reinmar** [comment]
Cause: "Duplication is better than the wrong abstraction".

**Comment by Reinmar** [comment]
Also, @panr mentioned that the implementation of this functionality differs slightly between commands and plugins (no refresh() in plugins) which is exactly the reason why it may be easier to duplicate the code.

**Comment by jodator** [comment]
Actually, it makes more sense as you mentioned. Do we want an interface for this? Probably we do as I remember the usage of this. Both Plugin and Command would implement the `ForceDisableable` interfaces (doh, the name...) in this case.

File: src/texttransformation.js [hunk 1]
```diff
@@ -92,12 +92,44 @@ export default class TextTransformation extends Plugin {
...
+
+		/**
+		 * Holds a set of active {@link module:typing/textwatcher~TextWatcher}
+		 *
+		 * @type {Set.<TextWatcher>}
+		 * @private
+		 */
+		this._watchersStack = new Set();
+
+		this.on( 'change:isEnabled', () => {
```
Comment (jodator) [inline src/texttransformation.js]: And I think that being enabled/disabled is also a property of a text watcher. It makes sense to enable/disable a text watcher while holding its instance. As we hae the first case others might also need this.

This way we will simplify this class by binding `watcher.bind( 'isEnabled' ).to.( this );`. But the `TextWatcher` class will need to be updated.

Comment (jodator) [reply src/texttransformation.js]: Also, I think the `watcherStack` shouldn't be required then.

File: src/textwatcher.js [hunk 2]
```diff
@@ -43,7 +43,7 @@ export default class TextWatcher {
 		const model = this.model;
```
Comment (jodator) [inline src/textwatcher.js]: Basically I think that adding an observable `isEnabled` property to this class will make easier to start/stop listening.

File: src/texttransformation.js [hunk 3]
```diff
@@ -92,12 +92,44 @@ export default class TextTransformation extends Plugin {
...
+			this.isEnabled ? this._enableTransformationWatchers() : this._disableTransformationWatchers();
+		} );
+
+		this._enableTransformationWatchers();
+
+		this.listenTo( modelDocument, 'change:data', () => {
```
[the remark this answers is not in the document]
Comment (panr) [reply src/texttransformation.js]: Oh... there will be sooooo many state updates 😓

File: tests/texttransformation.js [hunk 4]
```diff
@@ -40,6 +41,53 @@ describe( 'Text transformation feature', () => {
...
+			} );
+		} );
+
+		it( 'should be enabled after initialization', () => {
+			plugin.init();
+
+			expect( plugin.isEnabled ).to.be.true;
+		} );
+
+		it( 'should initialize watchers after initialization', () => {
```
[the remark this answers is not in the document]
Comment (panr) [reply tests/texttransformation.js]: Correct, that's why I spent too much time on this, trying to figure out the best way to handle this case, but yeah... I'm fine with removing this 👌

File: tests/texttransformation.js [hunk 5]
```diff
@@ -40,6 +41,53 @@ describe( 'Text transformation feature', () => {
...
+		let plugin;
+
+		beforeEach( () => {
+			return createEditorInstance().then( () => {
+				plugin = editor.plugins.get( TextTransformation );
+			} );
+		} );
+
+		it( 'should be enabled after initialization', () => {
+			plugin.init();
```
[the remark this answers is not in the document]
Comment (panr) [reply tests/texttransformation.js]: 👌

File: src/texttransformation.js (hunk 1 above)
Comment (panr) [reply src/texttransformation.js]: Changed, hope I've done it right ;-)

File: src/textwatcher.js (hunk 2 above)
Comment (panr) [reply src/textwatcher.js]: 👍

**Comment by panr** [comment]
There's a little commit madness. Should I squash some of them?

File: src/textwatcher.js [hunk 6]
```diff
@@ -31,6 +32,8 @@ export default class TextWatcher {
 		this.testCallback = testCallback;
 		this.hasMatch = false;
 
+		this.set( 'isEnabled', true );
```
Comment (jodator) [inline src/textwatcher.js]: Docs are needed.

Comment (jodator) [reply src/textwatcher.js]: Also AFAIR the:
```
 * @mixes module:utils/observablemixin~ObservableMixin
```
is needed for docs generation.

File: src/textwatcher.js [hunk 7]
```diff
@@ -127,4 +140,4 @@ export default class TextWatcher {
 	}
 }
 
-mix( TextWatcher, EmitterMixin );
+mix( TextWatcher, EmitterMixin, ObservableMixin );
```
Comment (jodator) [inline src/textwatcher.js]: The `ObservableMixin` is enough. It already mixes `EmitterMixin`.

File: tests/textwatcher.js [hunk 8]
```diff
@@ -40,12 +40,25 @@ describe( 'TextWatcher', () => {
 
 	afterEach( () => {
 		sinon.restore();
+		watcher.off();
```
Comment (jodator) [inline tests/textwatcher.js]: Shouldn't be necessary AFAICS.

After each test, a new instance is created and old references should be released.

File: tests/textwatcher.js [hunk 9]
```diff
@@ -188,6 +266,31 @@ describe( 'TextWatcher', () => {
...
+			} );
+
+			sinon.assert.notCalled( testCallbackStub );
+			sinon.assert.notCalled( matchedDataSpy );
+			sinon.assert.notCalled( matchedSelectionSpy );
+			sinon.assert.notCalled( unmatchedSpy );
+		} );
 	} );
 } );
-
```
Comment (jodator) [inline tests/textwatcher.js]: :+1:
